### ISSUE TITLE
clf() and cla() throws warnings with shared loglog plots.

### ISSUE THREAD

**Issue by fredrik-1** [matplotlib/matplotlib#9970]
``` 
import matplotlib.pyplot as plt

fig=plt.figure(1)
ax1=fig.add_subplot(1,2,1)
ax1.loglog(range(10))
ax2=fig.add_subplot(1,2,2, sharex=ax1)
ax2.loglog(range(10))

fig.clf()
```
Throws a warning:
 UserWarning: Attempted to set non-positive xlimits for log-scale axis; invalid limits will be ignored.
  'Attempted to set non-positive xlimits for log-scale axis;  

ax1.cla() also throws a warning. 

**Matplotlib version**
2.1.0,  
python 3.6.3
ipython 6.1.0

**Comment by lkjell** [comment]
fig.clf() will do a ax1.cla(). So the real culpit is ax1.cla().
Because ax1 does not have a master/parent axes share it will default to 'linear' scale. And then set the data limit to [0,1]. But since ax1 share axes with ax2 which has log scale it will throw the warning.

So there are 2 problems here. 1) Why should fig.clf() clear axes as well. It should just clear the axes from the figure and not clear the axes itself.
2)  when a master axes clear should it revert all the other shared axes to linear scale as well.

**Comment by divenex** [comment]
This bug is still present in Matplotlib 2.2

**Comment by divenex** [comment]
This bug still persists in Matplotlib 3.3

**Comment by jklymak** [comment]
You are trying to plot 0 on a loglog plot so you get a warning.  I am not sure what the bug is.

**Comment by timhoffm** [comment]
> You are trying to plot 0 on a loglog plot so you get a warning. I am not sure what the bug is.

Yes, but probably other than you think :smile:. This is the minimal example I cam up with (no 0 data involved):

~~~
fig, (ax1, ax2) = plt.subplots(1, 2, sharex=True)
x = range(1, 10)
ax1.plot(x, x)
ax2.semilogx(x, x)
plt.clf()
~~~

Ingedients you need:
- at least one log scale on a shared axes
- `plt.clf()` without a prior `plt.show()`

I suspect this has to do something with 
1) an internal state not being updated (missing `show()`) so that we still have a 0 limit.
2) The interaction between shared axes during cleanup.

Given that this only happens when you clear a non-shown figure (that also needs certain further properties), I'd be fine with closing as "won't fix".

**Comment by anntzer** [comment]
Actually the warning shows up even if the figure is shown (e.g. if you type @timhoffm's lines one after the other at an IPython prompt).  AFAICT the root cause is that `cla()`'ing ax1 resets it to linear, but doesn't reset the scale on ax2, so we end up with ax1 using log and ax2 using linear scale, which seems contrary to the semantics of shared axes.

Also there's some funky asymmetry which makes `ax2.cla()` not do the same (it doesn't even revert the scale to linear), likely because the implementation in `cla` is
```python
        if self._sharex is not None:
            self.sharex(self._sharex)
        else:
            self.xaxis._set_scale('linear')
            try:
                self.set_xlim(0, 1)
            except TypeError:
                pass
```
so things behave differently between the master axes (which has `self._sharex=None`) and the follower axes (which as `self._sharex = master`); at least that seems like something to be fixed...

**Comment by Fernal73** [comment]
```import warnings
warnings.filterwarnings("ignore", message="Attempted to set non-positive left xlim on a log-scaled axis.\nInvalid limit will be ignored.")```

Ignore the warning as above.

**Comment by billtubbs** [comment]
It would be great to have this fixed!  It's raising warnings in our test script which is like this:

```
fig, axes = plt.subplots(2, 1, sharex=True)
axes[0].loglog([1e-1, 1e0, 1e1], [10, 100, 1000])
axes[1].loglog([1e-1, 1e0, 1e1], [10, 100, 1000])
# ... do some tests here ...
fig.clear()
```

```text
<ipython-input-19-186c7546926a>:1: UserWarning: Attempted to set non-positive left xlim on a log-scaled axis.
Invalid limit will be ignored.
  fig.clear()
```

**Comment by gjcoram** [comment]
> ```python
> import warnings
> warnings.filterwarnings("ignore", message="Attempted to set non-positive left xlim on a log-scaled axis.\nInvalid limit will be ignored.")
> ```
> 
> Ignore the warning as above.

I just added
plt.xcale('linear')
right before the cla() call, to ensure the x-axis is linear. Fewer characters. :)

**Comment by pinradaThan** [comment]
Hello! Seems like this issue has not been fixed yet. My friend and I are first-time contributors and would like to take a jab at this.

**Comment by saumyacoder1709** [comment]
I would like to work on this issue.

**Comment by tacaswell** [comment]
https://github.com/matplotlib/matplotlib/pull/30843 may have a partial fix (re-ordering the way state is set) but has some other changes that are not correct.

Shared axis can not logically have different scales (otherwise they are not the same!) so I suspect the complete fix is in making sure `cla` correctly resets the linked axis as well.
